register.js

```diff
@@ -90,21 +101,39 @@
 	}
 
 	function wrapInZone(object, property, fn, global) {
-		if(fn) {
-			fn = fn(object[property]);
-		} else {
-			fn = object[property];
-		}
-		var wrappedFn = function(){
+		var wrappedFn = function() {
 			var Zone = global.CanZone;
-			if(typeof Zone !== "undefined" && !!Zone.current) {
+			if (typeof Zone !== "undefined" && !!Zone.current) {
 				return Zone.tasks[property](fn, Zone).apply(this, arguments);
 			}
 
 			return fn.apply(this, arguments);
 		};
+
+		var descriptor = Object.getOwnPropertyDescriptor(object, property) || {};
+
+		if (isGlobalEventHandler(property)) {
+			fn = descriptor.set;
+			descriptor.set = wrappedFn;
+		} else {
+			fn = fn ? fn(object[property]) : object[property];
+			descriptor.value = wrappedFn;
+		}
+
+		Object.defineProperty(object, property, descriptor);
 		wrappedFn.zoneWrapped = true;
-		object[property] = wrappedFn;
+	}
+
+	function isGlobalEventHandler(property) {
+		return property.substr(0, 2) === "on";
+	}
+
+	// Returns an array of global event handlers names
+	// https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers
+	function getGlobalEventHandlersNames() {
```

Comment:
I did not put this in a shared module b/c for some reason I though we don't want any `require` in this module. Is that right?

Reply:
Yeah, it's weird.